aws/resource_aws_globalaccelerator_listener.go

```diff
@@ -0,0 +1,270 @@
+		opts.ClientAffinity = aws.String(v.(string))
+	}
+
+	log.Printf("[DEBUG] Create Global Accelerator listener: %s", opts)
+
+	resp, err := conn.CreateListener(opts)
+	if err != nil {
+		return fmt.Errorf("Error creating Global Accelerator listener: %s", err)
+	}
+
+	d.SetId(*resp.Listener.ListenerArn)
+
+	// Creating a listener triggers the accelerator to change status to InPending
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{globalaccelerator.AcceleratorStatusInProgress},
+		Target:  []string{globalaccelerator.AcceleratorStatusDeployed},
+		Refresh: resourceAwsGlobalAcceleratorAcceleratorStateRefreshFunc(conn, d.Get("accelerator_arn").(string)),
+		Timeout: d.Timeout(schema.TimeoutCreate),
+	}
+
+	log.Printf("[DEBUG] Waiting for Global Accelerator listener (%s) availability", d.Id())
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("Error waiting for Global Accelerator listener (%s) availability: %s", d.Id(), err)
+	}
+
+	return resourceAwsGlobalAcceleratorListenerRead(d, meta)
+}
+
+func resourceAwsGlobalAcceleratorListenerRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).globalacceleratorconn
+
+	listener, err := resourceAwsGlobalAcceleratorListenerRetrieve(conn, d.Id())
+
+	if err != nil {
+		return fmt.Errorf("Error reading Global Accelerator listener: %s", err)
+	}
+
+	if listener == nil {
+		log.Printf("[WARN] Global Accelerator listener (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	acceleratorArn, err := resourceAwsGlobalAcceleratorListenerParseAcceleratorArn(d.Id())
+
+	if err != nil {
+		return err
+	}
+
+	d.Set("accelerator_arn", acceleratorArn)
+	d.Set("client_affinity", listener.ClientAffinity)
+	d.Set("protocol", listener.Protocol)
+	d.Set("port_range", resourceAwsGlobalAcceleratorListenerFlattenPortRanges(listener.PortRanges))
+
+	return nil
+}
+
+func resourceAwsGlobalAcceleratorListenerParseAcceleratorArn(listenerArn string) (string, error) {
+	parts := strings.Split(listenerArn, "/")
+	if len(parts) < 4 {
+		return "", fmt.Errorf("Unable to parse accelerator ARN from %s", listenerArn)
+	}
+	return strings.Join(parts[0:2], "/"), nil
+}
+
+func resourceAwsGlobalAcceleratorListenerExpandPortRanges(portRanges []interface{}) []*globalaccelerator.PortRange {
+	out := make([]*globalaccelerator.PortRange, len(portRanges))
+
+	for i, raw := range portRanges {
+		portRange := raw.(map[string]interface{})
+		m := globalaccelerator.PortRange{}
+
+		m.FromPort = aws.Int64(int64(portRange["from_port"].(int)))
+		m.ToPort = aws.Int64(int64(portRange["to_port"].(int)))
+
+		out[i] = &m
+	}
+
+	return out
+}
+
+func resourceAwsGlobalAcceleratorListenerFlattenPortRanges(portRanges []*globalaccelerator.PortRange) []interface{} {
+	out := make([]interface{}, len(portRanges))
+
+	for i, portRange := range portRanges {
+		m := make(map[string]interface{})
+
+		m["from_port"] = *portRange.FromPort
```

Comment:
To prevent potential panics we should use the AWS Go SDK provided helpers with FromPort and ToPort:

```go
		m["from_port"] = aws.Int64Value(portRange.FromPort)
		m["to_port"] = aws.Int64Value(portRange.ToPort)
```